WalletWasabi/BestEffortEndpointConnector.cs

```diff
@@ -0,0 +1,120 @@
+using NBitcoin;
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using NBitcoin.Protocol.Connectors;
+using NBitcoin.Socks;
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi
+{
+	public class BestEffortEndpointConnector : IEnpointConnector
+	{
+		public BestEffortEndpointConnector(long maxNonOnionConnectionCount)
+			: this(new EffortState(maxNonOnionConnectionCount))
+		{
+		}
+
+		private BestEffortEndpointConnector(EffortState state)
+		{
+			State = state;
+		}
+
+		public EffortState State { get; private set; }
+
+		public void UpdateConnectedNodesCounter(int connectedNodes)
+		{
+			State.ConnectedNodesCount = connectedNodes;
+		}
+
+		public IEnpointConnector Clone()
+		{
+			return new BestEffortEndpointConnector(State);
+		}
+
+		public virtual async Task ConnectSocket(Socket socket, EndPoint endpoint, NodeConnectionParameters nodeConnectionParameters, CancellationToken cancellationToken)
+		{
+			var isTor = endpoint.IsTor();
+
+			var socksSettings = nodeConnectionParameters.TemplateBehaviors.Find<SocksSettingsBehavior>();
+			var socketEndpoint = endpoint;
+			var useSocks = isTor || socksSettings?.OnlyForOnionHosts is false;
+			if (useSocks)
+			{
+				if (socksSettings?.SocksEndpoint == null)
+				{
+					throw new InvalidOperationException("SocksSettingsBehavior.SocksEndpoint is not set but the connection is expecting using socks proxy");
+				}
+				if (!isTor && State.AllowOnlyTorEndpoints)
+				{
+					throw new InvalidOperationException($"The Endpoint connector is configured to allow only Tor endpoints and the '{endpoint}' enpoint is not one");
+				}
+
+				socketEndpoint = socksSettings.SocksEndpoint;
+			}
+
+			if (socketEndpoint is IPEndPoint mappedv4 && mappedv4.Address.IsIPv4MappedToIPv6Ex())
+			{
+				socketEndpoint = new IPEndPoint(mappedv4.Address.MapToIPv4Ex(), mappedv4.Port);
+			}
+
+			await socket.ConnectAsync(socketEndpoint).WithAwaitCancellationAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
@lontivero this PR is still under construction but. I had to replace your line with this
https://github.com/zkSNACKs/WalletWasabi/pull/5024/files?file-filters%5B%5D=.cs&hide-deleted-files=true#diff-7af935310c99f4b468057fec34fbeb37b84c1d8d6afca38caa6d4a490f6f9c4aR65

In the backport branch we are still using .NET Core 3.1 that does not have a cancellation. Is this good?

Reply:
Yes @molnard, it is that or just remove the `WithAwaitCancellationAsync` and call `cancel.ThrowIfCancellationRequested` before, it is not the same thing I know.

`WithAwaitCancellationAsync` is no good, it is just that there were no other way to cancel operations that didn't receive a cancellation token but the truth is that we are not cancelling the task at all, we just stop waiting for the completion so, the tasks continue executing and doing things anyway. I am happy we can finally remove them all (or at least most of them)

Reply:
ACK